compiler/qsc/src/interpret/stateful.rs

```diff
@@ -438,111 +490,54 @@ impl Interpreter {
-        if self.target != TargetProfile::Base {
-            return Err(vec![Error(ErrorKind::TargetMismatch)]);
-        }
-
-        let (items, stmts) = self.compile_incremental(|c, p| c.compile_expr(p, "<entry>", expr))?;
-
-        let mut ret = Ok(String::new());
-
-        for item in items {
-            match item.kind {
-                qsc_hir::hir::ItemKind::Callable(callable) => {
-                    let callable = self.lower_callable_decl(&callable);
-
-                    self.callables
-                        .insert(qsc_eval::lower::lower_local_item_id(item.id), callable);
-                }
-                qsc_hir::hir::ItemKind::Namespace(..) => {}
-                qsc_hir::hir::ItemKind::Ty(..) => {
-                    self.udts
-                        .insert(qsc_eval::lower::lower_local_item_id(item.id));
-                }
+        let mut results: Vec<InterpretResult> = Vec::new();
+        for i in 0..shots {
+            for stmt_id in &stmt_ids {
+                results.push(
+                    match eval_stmt(
+                        *stmt_id,
+                        &globals,
+                        &mut Env::with_empty_scope(),
+                        sim,
+                        self.package,
+                        receiver,
+                    ) {
+                        Ok(value) => Ok(value),
+                        Err((error, call_stack)) => {
+                            let stack_trace = if call_stack.is_empty() {
+                                None
+                            } else {
+                                Some(self.render_call_stack(call_stack, &error))
+                            };
+
+                            Err(vec![Error(
+                                error::from_eval(error, &self.store, stack_trace).into(),
+                            )])
+                        }
+                    },
+                );
             }
-        }
-
-        for stmt in stmts {
-            let stmt_id = self.lower_stmt(&stmt);
-            let globals = Lookup {
-                fir_store: &self.fir_store,
-                package: self.package,
-                udts: &self.udts,
-                callables: &self.callables,
-            };
 
-            ret = generate_qir_for_stmt(stmt_id, &globals, &mut self.env, self.package).map_err(
-                |(error, call_stack)| {
-                    let stack_trace = if call_stack.is_empty() {
-                        None
-                    } else {
-                        Some(self.render_call_stack(call_stack, &error))
-                    };
-
-                    vec![Error(
-                        error::from_eval(error, &self.store, stack_trace).into(),
-                    )]
-                },
-            );
+            if i != 0 {
+                // If running more than one shot, re-initialize the simulator to start the next shot
+                // from a clean state.
+                sim.reinit();
```

Comment:
Perhaps instead of requiring a `reinit` method, we can have this function take a function that creates a new simulator instance? I kind of liked the old approach of using `SparseSim::new()`. IMO, it makes it clearer that absolutely no state will be shared between instances.

Reply:
That approach would work for `run` but for other callers that need to interact with the simulator after execution it gets harder. The function would need to return an owned instance, which means the caller wouldn't get to keep it after execution. The `qirgen` function, for example, needs to call `finish` on the simulator get the constructed QIR out.